Sources/CacheAdvance/CacheAdvance.swift

```diff
@@ -172,10 +170,28 @@ public final class CacheAdvance<T: Codable> {
         try header.checkFile()
 
         var messages = [T]()
-        while let encodedMessage = try reader.nextEncodedMessage() {
-            messages.append(try decoder.decode(T.self, from: encodedMessage))
-        }
+        // There is only one range: | `offsetInFileOfOldestMessage` -> `offsetInFileAtEndOfNewestMessage`|
+        if reader.offsetInFileOfOldestMessage < reader.offsetInFileAtEndOfNewestMessage {
+            for encodedMessage in try reader.encodedMessagesFromOffset(reader.offsetInFileOfOldestMessage,
+                                                                       endOffset: reader.offsetInFileAtEndOfNewestMessage) {
+                messages.append(try decoder.decode(T.self, from: encodedMessage))
+            }
+        } else {
+            // In this case, the messages could be split to two ranges
+            // | First Range | (GAP: ignore) | Second Range |
```

Comment:
Does there _need_ to be a gap? I think technically the first range could be contiguous with the second range. If it's true that the ranges can be contiguous we should indicate that the gap is optional, for extra clarity.

Reply:
I'd need to dig in more, but we have a unit test where we documented that we expect there to be a gap and we do not believe the ranges can be contiguous:
https://github.com/dfed/CacheAdvance/blob/3a1ed6f99c2a870dc09c3a8144a58f915757a569/Tests/CacheAdvanceTests/CacheAdvanceTests.swift#L350-L389

Reply:
The gap is generated when the new message offset override the old message. Usually the message length of new is not the same as the old message to be rewrote. Then there probably is a gap.

Reply:
`test_append_inAnOverwritingCacheEvictsMinimumPossibleNumberOfMessages()` proves that there must be a gap. I'm resolving this comment thread.